js/commonjs/rewrite_tests_for_commonjs.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview Utility to translate test files to CommonJS imports.
+ *
+ * This is a somewhat hacky tool designed to do one very specific thing.
+ * All of the test files in *_test.js are written with Closure-style
+ * imports (goog.require()).  This works great for running the tests
+ * against Closure-style generated code, but we also want to run the
+ * tests against CommonJS-style generated code without having to fork
+ * the tests.
+ *
+ * Closure-style imports import each individual type by name.  This is
+ * very different than CommonJS imports which are by file.  So we put
+ * special comments in these tests like:
+ *
+ * // CommonJS-LoadFromFile: test_pb
+ * goog.require('proto.jspb.test.CloneExtension');
+ * goog.require('proto.jspb.test.Complex');
+ * goog.require('proto.jspb.test.DefaultValues');
+ *
+ * This script parses that special comment and uses it to generate proper
+ * CommonJS require() statements so that the tests can run and pass using
+ * CommonJS imports.
+ */
+
+var lineReader = require('readline').createInterface({
+  input: process.stdin,
+  output: process.stdout
+});
+
+function tryStripPrefix(str, prefix) {
+  if (str.lastIndexOf(prefix) !== 0) {
+    throw "String: " + str + " didn't start with: " + prefix;
+  }
+  return str.substr(prefix.length);
+}
+
+var module = null;
+var pkg = null;
+lineReader.on('line', function(line) {
+  var is_require = line.match(/goog\.require\('([^']*)'\)/);
+  var is_loadfromfile = line.match(/CommonJS-LoadFromFile: ([^ ]*) (.*)/);
+  var is_settestonly = line.match(/goog.setTestOnly()/);
+  if (is_settestonly) {
+    // Remove this line.
+  } else if (is_require) {
+    if (module) {  // Skip goog.require() lines before the first directive.
+      var full_sym = is_require[1];
+      var sym = tryStripPrefix(full_sym, pkg);
+      console.log("google_protobuf.exportSymbol('" + full_sym + "', " + module + sym + ', global);');
+    }
+  } else if (is_loadfromfile) {
+    if (!module) {
+      console.log("var google_protobuf = require('google-protobuf');");
+      console.log("var asserts = require('closure_asserts_commonjs');");
+      console.log("var global = Function('return this')();");
+      console.log("");
+      console.log("// Bring asserts into the global namespace.");
+      console.log("google_protobuf.object.extend(global, asserts);");
```

Comment:
`export_asserts.js` takes symbols that are global inside that module and exports them. That allows a test to say:

```javascript
var asserts = require('closure_asserts_commonjs');

// Now this symbol is available:
asserts.assertObjectEquals(1, 1);
```

But the tests do not call the asserts like `asserts.assertObjectEquals()`, they just say `assertObjectEquals()`. So we need to take the keys in the `asserts` object and put them in the global namespace of the test.

Put another way, both are necessary because the global namespace of the module is not the same as the global namespace of the test. The module can't put things in the test's global namespace. CommonJS modules are more hygienic than Closure modules in this way.